File: packages/python-sdk/src/mogbank/marketplace.py
"""Marketplace module — list services and manage escrow payments."""

from __future__ import annotations
from typing import Optional, List, Dict, Any
from dataclasses import dataclass

from mogbank.client import MogBankClient
from mogbank.types import Service, Escrow, Currency


@dataclass
class ServiceListingParams:
    agent_id: str
    name: str
    description: str
    price: float
    currency: Currency = Currency.USDC


@dataclass
class EscrowParams:
    buyer_agent_id: str
    seller_agent_id: str
    service_id: str
    amount: float
    currency: Currency = Currency.USDC
    buyer_wallet_id: str = ""
    seller_wallet_id: str = ""
    timeout_hours: int = 72


@dataclass
class EscrowResponse:
    escrow: Escrow
    transaction_id: str


class MarketplaceModule:
    """"""

    def __init__(self, client: MogBankClient):
        self.client = client

    def list_services(
        self,
        status: Optional[str] = None,
        min_price: Optional[float] = None,
        max_price: Optional[float] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> List[Service]:
        params = []
        if status:
            params.append(f"status={status}")
        if min_price is not None:
            params.append(f"min_price={min_price}")
        if max_price is not None:
            params.append(f"max_price={max_price}")
        params.extend([f"limit={limit}", f"offset={offset}"])
        query = "&".join(params)

        result = self.client.request(
            "GET", f"/api/v1/marketplace/services?{query}"
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Failed to list services")
        return [Service(**s) for s in result.data]

    def get_service(self, service_id: str) -> Service:
        result = self.client.request(
            "GET", f"/api/v1/marketplace/services/{service_id}"
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Service not found")
        return Service(**result.data)

    def create_service(self, params: ServiceListingParams) -> Service:
        import uuid

        result = self.client.request(
            "POST",
            "/api/v1/marketplace/services",
            {
                "agent_id": params.agent_id,
                "name": params.name,
                "description": params.description,
                "price": params.price,
                "currency": params.currency.value,
            },
            idempotency_key=f"svc_{params.agent_id}_{uuid.uuid4().hex[:8]}",
            sign_with_ed25519=True,
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Service creation failed")
        return Service(**result.data)

    def create_escrow(self, params: EscrowParams) -> EscrowResponse:
        import uuid

        result = self.client.request(
            "POST",
            "/api/v1/marketplace/escrow",
            {
                "buyer_agent_id": params.buyer_agent_id,
                "seller_agent_id": params.seller_agent_id,
                "service_id": params.service_id,
                "amount": params.amount,
                "currency": params.currency.value,
                "buyer_wallet_id": params.buyer_wallet_id,
                "seller_wallet_id": params.seller_wallet_id,
                "timeout_hours": params.timeout_hours,
            },
            idempotency_key=f"escrow_{params.service_id}_{uuid.uuid4().hex[:8]}",
            sign_with_ed25519=True,
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Escrow creation failed")
        data = result.data
        return EscrowResponse(
            escrow=Escrow(**data["escrow"]),
            transaction_id=data["transaction_id"],
        )

    def release_escrow(self, escrow_id: str) -> Escrow:
        result = self.client.request(
            "POST",
            f"/api/v1/marketplace/escrow/{escrow_id}/release",
            sign_with_ed25519=True,
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Escrow release failed")
        return Escrow(**result.data)

    def refund_escrow(self, escrow_id: str) -> Escrow:
        result = self.client.request(
            "POST",
            f"/api/v1/marketplace/escrow/{escrow_id}/refund",
            sign_with_ed25519=True,
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Escrow refund failed")
        return Escrow(**result.data)

    def get_escrow(self, escrow_id: str) -> Escrow:
        result = self.client.request(
            "GET", f"/api/v1/marketplace/escrow/{escrow_id}"
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Escrow not found")
        return Escrow(**result.data)

    def list_escrows(
        self, agent_id: Optional[str] = None, status: Optional[str] = None
    ) -> List[Escrow]:
        params = []
        if agent_id:
            params.append(f"agent_id={agent_id}")
        if status:
            params.append(f"status={status}")
        query = "&".join(params)

        result = self.client.request(
            "GET", f"/api/v1/marketplace/escrow?{query}"
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Failed to list escrows")
        return [Escrow(**e) for e in result.data]

File: packages/typescript-sdk/src/marketplace.ts
/**
 * Marketplace module — list, discover, and purchase AI agent services.
 * Supports escrow-based payments between agents.
 */
import { MogBankClient } from './client';
import { Service, Escrow, ApiResponse, Currency } from './types';

export type ServiceListing = Service;

export interface ServiceListingParams {
  seller_agent_id: string;
  name: string;
  description: string;
  price: number;
  currency?: Currency;
}

export interface EscrowParams {
  buyer_agent_id: string;
  seller_agent_id: string;
  service_id: string;
  amount: number;
  currency?: Currency;
  /** Timeout in hours before escrow auto-refunds (default: 72) */
  timeout_hours?: number;
}

export interface EscrowResponse {
  escrow: Escrow;
  transaction_id: string;
}

export class MarketplaceModule {
  constructor(private client: MogBankClient) {}

  /**
   * List a new service for sale on the marketplace.
   */
  async listService(params: ServiceListingParams): Promise<Service> {
    const result = await this.client.request<ApiResponse<Service>>(
      'POST',
      '/api/v1/marketplace/services',
      params,
      {
        idempotencyKey: `svc_${params.seller_agent_id}_${Date.now()}`,
        signWithEd25519: true,
      }
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Failed to list service');
    }
    return result.data;
  }

  /**
   * Search for available services.
   */
  async searchServices(query?: {
    name?: string;
    seller_agent_id?: string;
    min_price?: number;
    max_price?: number;
    status?: 'active' | 'inactive';
    limit?: number;
    offset?: number;
  }): Promise<Service[]> {
    const params = new URLSearchParams();
    if (query?.name) params.set('name', query.name);
    if (query?.seller_agent_id) params.set('seller_agent_id', query.seller_agent_id);
    if (query?.min_price !== undefined) params.set('min_price', String(query.min_price));
    if (query?.max_price !== undefined) params.set('max_price', String(query.max_price));
    if (query?.status) params.set('status', query.status);
    if (query?.limit) params.set('limit', String(query.limit));
    if (query?.offset) params.set('offset', String(query.offset));

    const result = await this.client.request<ApiResponse<Service[]>>(
      'GET',
      `/api/v1/marketplace/services?${params.toString()}`
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Failed to search services');
    }
    return result.data;
  }

  /**
   * Get a specific service listing.
   */
  async getService(serviceId: string): Promise<Service> {
    const result = await this.client.request<ApiResponse<Service>>(
      'GET',
      `/api/v1/marketplace/services/${serviceId}`
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Service not found');
    }
    return result.data;
  }

  /**
   * Update a service listing (by seller).
   */
  async updateService(
    serviceId: string,
    updates: {
      name?: string;
      description?: string;
      price?: number;
      status?: 'active' | 'inactive' | 'deprecated';
    }
  ): Promise<Service> {
    const result = await this.client.request<ApiResponse<Service>>(
      'PATCH',
      `/api/v1/marketplace/services/${serviceId}`,
      updates,
      {
        idempotencyKey: `svc_upd_${serviceId}_${Date.now()}`,
        signWithEd25519: true,
      }
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Failed to update service');
    }
    return result.data;
  }

  /**
   * Lock funds into escrow to purchase a service.
   * 
   * This creates an escrow agreement between buyer and seller.
   * Funds are held until the buyer releases them or the timeout expires.
   */
  async createEscrow(params: EscrowParams): Promise<EscrowResponse> {
    const result = await this.client.request<ApiResponse<EscrowResponse>>(
      'POST',
      '/api/v1/marketplace/escrow',
      params,
      {
        idempotencyKey: `esc_${params.buyer_agent_id}_${params.service_id}_${Date.now()}`,
        signWithEd25519: true,
      }
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Failed to create escrow');
    }
    return result.data;
  }

  /**
   * Release escrow funds to the seller (buyer confirms delivery).
   */
  async releaseEscrow(escrowId: string): Promise<EscrowResponse> {
    const result = await this.client.request<ApiResponse<EscrowResponse>>(
      'POST',
      `/api/v1/marketplace/escrow/${escrowId}/release`,
      {},
      {
        idempotencyKey: `esc_rel_${escrowId}_${Date.now()}`,
        signWithEd25519: true,
      }
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Failed to release escrow');
    }
    return result.data;
  }

  /**
   * Refund escrow back to buyer (e.g., dispute resolved, timeout).
   */
  async refundEscrow(escrowId: string): Promise<EscrowResponse> {
    const result = await this.client.request<ApiResponse<EscrowResponse>>(
      'POST',
      `/api/v1/marketplace/escrow/${escrowId}/refund`,
      {},
      {
        idempotencyKey: `esc_ref_${escrowId}_${Date.now()}`,
        signWithEd25519: true,
      }
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Failed to refund escrow');
    }
    return result.data;
  }

  /**
   * Get escrow details.
   */
  async getEscrow(escrowId: string): Promise<Escrow> {
    const result = await this.client.request<ApiResponse<Escrow>>(
      'GET',
      `/api/v1/marketplace/escrow/${escrowId}`
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Escrow not found');
    }
    return result.data;
  }
}